### DISCUSSION TITLE
Laravel Resent Password Link to Admins

### DISCUSSION THREAD

**Discussion by KarimDe1** [community/community#138719] [Programming Help]
### Select Topic Area

Question

### Body

**Laravel 8** , I'm trying to send a password reset link specifically to admin users. 
I've already created a separate guard for the admins, but I'm not sure how to properly configure the password reset for this group.

**Comment by halitworks** [comment]
To send a password reset link specifically to admin users in Laravel 8, you've already set up a separate guard for admins, which is a good start. Now, you need to configure the password reset functionality for this guard. Here’s how you can do it:

### Steps to Configure Password Reset for Admins:

#### 1. **Configure `auth.php`**
In `config/auth.php`, make sure you have a separate guard for admins, and also define the `passwords` section for the admin.

```php
'guards' => [
    'web' => [
        'driver' => 'session',
        'provider' => 'users',
    ],
    
    'admin' => [
        'driver' => 'session',
        'provider' => 'admins',
    ],
],

'providers' => [
    'users' => [
        'driver' => 'eloquent',
        'model' => App\Models\User::class,
    ],

    'admins' => [
        'driver' => 'eloquent',
        'model' => App\Models\Admin::class, // Define your admin model
    ],
],

'passwords' => [
    'users' => [
        'provider' => 'users',
        'table' => 'password_resets',
        'expire' => 60,
        'throttle' => 60,
    ],

    'admins' => [
        'provider' => 'admins',  // The admin provider defined above
        'table' => 'password_resets', // Can use the same table or a separate one
        'expire' => 60,
        'throttle' => 60,
    ],
],
```

#### 2. **Admin Password Reset Controller**
Create a custom password reset controller for admin users. This controller should extend Laravel’s default password reset controller, but use the `admin` guard and configuration.

```php
php artisan make:controller AdminForgotPasswordController --namespace=Auth
```

In the controller:

```php
namespace App\Http\Controllers\Auth;

use App\Http\Controllers\Controller;
use Illuminate\Http\Request;
use Password;
use Illuminate\Support\Facades\Password;

class AdminForgotPasswordController extends Controller
{
    public function showLinkRequestForm()
    {
        return view('auth.passwords.email-admin'); // Create a view for admin reset link
    }

    public function sendResetLinkEmail(Request $request)
    {
        $request->validate(['email' => 'required|email']);

        // Use the admin password broker
        $status = Password::broker('admins')->sendResetLink(
            $request->only('email')
        );

        return $status === Password::RESET_LINK_SENT
            ? back()->with(['status' => __($status)])
            : back()->withErrors(['email' => __($status)]);
    }
}
```

#### 3. **Admin Password Reset Broker**
In your `sendResetLinkEmail` method, use `Password::broker('admins')` to make sure you're sending the reset link via the admin guard.

#### 4. **Admin Routes**
Define custom routes for the admin password reset functionality in `routes/web.php`:

```php
// Admin password reset routes
Route::get('admin/password/reset', [AdminForgotPasswordController::class, 'showLinkRequestForm'])->name('admin.password.request');
Route::post('admin/password/email', [AdminForgotPasswordController::class, 'sendResetLinkEmail'])->name('admin.password.email');
```

#### 5. **Admin Password Reset View**
Create a custom view for the admin password reset form (`resources/views/auth/passwords/email-admin.blade.php`).

```php
<form method="POST" action="{{ route('admin.password.email') }}">
    @csrf
    <input type="email" name="email" required autofocus>
    <button type="submit">Send Password Reset Link</button>
</form>
```

#### 6. **Admin Reset Password Controller**
You’ll also need a controller to handle the actual password reset process for admins. You can create it similarly to the forgot password controller.

```php
php artisan make:controller AdminResetPasswordController --namespace=Auth
```

In the controller, handle the reset process:

```php
namespace App\Http\Controllers\Auth;

use App\Http\Controllers\Controller;
use Illuminate\Foundation\Auth\ResetsPasswords;
use Illuminate\Http\Request;
use Password;

class AdminResetPasswordController extends Controller
{
    use ResetsPasswords;

    protected function broker()
    {
        return Password::broker('admins'); // Use the admin broker
    }

    protected function guard()
    {
        return auth()->guard('admin'); // Use the admin guard
    }
}
```

#### 7. **Testing the Flow**
- **Visit the reset link form:** `/admin/password/reset`
- **Send a reset link:** After entering the admin email, the system will send a reset link using the `admins` provider.

By following these steps, you'll be able to send a password reset link specifically to your admin users while maintaining separation from the regular user functionality.

**Comment by CoderZule** [comment]
in controller  ResetPassword function make sure to use  Laravel's password broker: 
$status = Password::broker('admins')->sendResetLink($request->only('email'));

### config/auth.php

'passwords' => [
    'users' => [
        'provider' => 'users',
        'table' => 'password_resets',
        'expire' => 60,
    ],

    'admins' => [
        'provider' => 'admins',
        'table' => 'password_resets',
        'expire' => 60,
    ],
],

'guards' => [
    'admin' => [
        'driver' => 'session',
        'provider' => 'admins',
    ],
],

'providers' => [
    'admins' => [
        'driver' => 'eloquent',
        'model' => App\Models\Admin::class,
    ],
],

**Comment by LiteBrite82** [comment]
Thanks for posting in the GitHub Community, @KarimDe1 ! 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.

**Comment by ebndev** [comment]
Please see [this comment](https://github.com/orgs/community/discussions/138722#discussioncomment-10664488) for reference to why this discussion was closed and locked. Thanks!
